File: Data Structures/Structures/Stack/Stack.java
package Structures.Stack;

public class Stack<E> implements IStack<E> {
    // holds size of stack
    private int size;
    // holds data for the stack
    private Object[] arr;
    // holds position of top item
    private int top;

    /**
     * Constructor method for the stack. 
     *
     * @param s   the input array to sort
     */
    public Stack(int s) {
        // sets top to -1 & creates an array of size given by user
        this.size = s;
        this.arr = new Object[s];
        this.top = -1;
    }

    /**
     * Method to check if the stack is full. 
     *
     * @return      whether the stack is full
     */
    public boolean isFull() {
        return top == size - 1;
    }

    /**
     * Method to check if the stack is empty. 
     *
     * @return      whether the stack is empty
     */
    public boolean isEmpty() {
        return top == -1;
    }

    /**
     * Method to add an item to the stack. 
     *
     * @param   e   element to add to the stack 
     * @return      whether the item could be added or not
     */
    public boolean push(E e) {
        // checks if the stack is not full; if there is space for an item
        if (!isFull()) {
            // increments top
            this.top ++;
            // adds item at top
            this.arr[this.top] = e;
            // returns true
            return true;
        }
        // returns false if there is no space in the stack
        return false;
    }

    /**
     * Method to remove an item from the top of the stack. 
     *
     * @return      whether the item could be removed or not
     */
    public E pop() {
        // checks if there is an item to pop
        if (!isEmpty()) {
            // gets the top item
            E item = (E) this.arr[this.top];
            // decrements the top pointer
            this.top --;
            // returns true
            return item;
        }
        // returns null if an item couldn't be removed
        return null;
    }

    /**
     * Method to return the item at the top of the stack. 
     *
     * @return      the top item
     */
    public E top() {
        // checks if there are any items in the stack
        if (!isEmpty()) {
            // returns the top item
            return (E) this.arr[this.top];
        }
        // returns null if no item was found
        return null;
    }

    // tests
    public static void main(String[] args) {
        Stack s = new Stack<>(5);
        s.push(1);
        s.push(2);
        s.push(3);
        s.push(4);
        s.push(5);
        s.push(6);
        System.out.println(s.top());
        System.out.println(s.pop());
        System.out.println(s.pop());
        System.out.println(s.pop());
        System.out.println(s.pop());
        System.out.println(s.pop());
        System.out.println(s.pop());
    }
}

File: Data Structures/Data Structures/Structures/Stack/stack.py
class stack:
    # constructor
    def __init__(self, size):
        # sets top to -1
        self.__top = -1
        # sets size var to size given by user
        self.__size = size
        # creates array of size given by user
        self.__data = [None] * self.__size
    
    # checks if the stack is full
    def isFull(self):
        return (self.__top == self.__size - 1)
    
    # checks if the stack is empty
    def isEmpty(self):
        return (self.__top == -1)
    
    # returns the item at the top
    def top(self):
        # checks if the stack is not empty
        if not stack.isEmpty(self):
            # returns the top item
            return self.__data[self.__top]
        # returns None if no items are in the stack
        return None

    # adds an item to the stack
    def push(self, item):
        # id the stack has space
        if not stack.isFull(self):
            # increment top
            self.__top += 1
            # add data at the top of the stack
            self.__data[self.__top] = item
            # return true
            return True
        # return false if item was not added
        return False

    # method to remove an item form the stack
    def pop(self):
        # checks if there is an item to remove
        if not stack.isEmpty(self):
            # gets the item
            item = self.__data[self.__top]
            # decrements the top
            self.top -= 1
            # returns the item
            return item
        # returns None if stack was empty
        return None

    # shows all content of the stack
    def toString(self):
        # if the stack has items
        if not stack.isEmpty(self):
            # var to hold the items
            string = ""
            # add each item with a space (" ") between them
            for i in range(self.__top + 1):
                string += str(self.__data[i]) + " "
            # return the string
            return string
        # return None if the stack was empty
        return None

# tests
s = stack(5)
print(s.toString())
print(s.pop())
print(s.isEmpty())
print(s.isFull())
print(s.push(1))
print(s.push(2))
print(s.push(3))
print(s.push(4))
print(s.push(5))
print("isFull? ", s.isFull())
print(s.push(6))
print(s.toString())
print(s.top())

